src/MixedMatrix.cpp

```diff
@@ -15,221 +15,311 @@
 
 /** @file
 
-    @brief Implements MixedMatrix object.
+    @brief MixedMatrix class
 */
 
-#ifndef __MIXEDMATRIX_IMPL_HPP__
-#define __MIXEDMATRIX_IMPL_HPP__
-
 #include "MixedMatrix.hpp"
-#include "mfem.hpp"
-#include "utilities.hpp"
-#include <memory>
-
-using std::unique_ptr;
 
 namespace smoothg
 {
 
-MixedMatrix::MixedMatrix(const mfem::SparseMatrix& vertex_edge,
-                         const mfem::Vector& weight,
-                         const mfem::SparseMatrix& w_block,
-                         const mfem::HypreParMatrix& edge_d_td,
-                         DistributeWeight dist_weight)
-    : edge_d_td_(&edge_d_td),
-      edge_td_d_(edge_d_td_->Transpose())
+MixedMatrix::MixedMatrix(const Graph& graph)
+    : edge_true_edge_(graph.edge_true_edge_),
+      D_local_(MakeLocalD(graph.edge_true_edge_, graph.vertex_edge_local_)),
+      W_local_(graph.W_local_),
+      elem_dof_(graph.vertex_edge_local_),
+      agg_vertexdof_(SparseIdentity(D_local_.Rows())),
+      face_facedof_(SparseIdentity(elem_dof_.Cols()))
 {
-    assert(edge_d_td);
-    assert(weight.Size() == vertex_edge.Width());
-    assert(edge_d_td_->Height() == vertex_edge.Width());
+    int num_vertices = D_local_.Rows();
+    int num_edges = D_local_.Cols();
+
+    M_elem_.resize(num_vertices);
 
-    mfem::Vector weight_cut;
+    SparseMatrix edge_vertex = D_local_.Transpose();
+    std::vector<double> weight_inv = graph.weight_local_;
+
+    for (auto& i : weight_inv)
+    {
+        assert(std::fabs(i) > 1e-12);
+        assert(i == i);
```

Comment:
This was something else I believe, but certainly we dont need it now XD

Reply:
This was to make sure is not `NaN`, though i will change to `std::isfinite` for a better check.

Reply:
Aw, I see, now it makes sense to me. Thanks for the explanation, I was wondering why there is this check.